### PR TITLE
[OV JS] Parameters validation, templates approach

### PR DESCRIPTION
**Pull request by vishniakov-nikolai**
### Details:
 - Validate input parameters of JS binding methods using template approach

### Tickets:
 - 137672

Based on: https://github.com/almilosz/openvino/tree/almilosz/arg_val

### CODE REVIEW

File: src/bindings/js/node/include/type_validation.hpp [hunk 1]
```diff
@@ -5,7 +5,142 @@
...
+template <typename Arg, typename Arg1, typename Arg2>
+std::vector<std::string> get_signature() {
+    std::vector<std::string> attributes;
+
+    attributes.push_back(get_attr_type<Arg>());
+    attributes.push_back(get_attr_type<Arg1>());
+    attributes.push_back(get_attr_type<Arg2>());
+
+    return attributes;
+};
```
Comment (t-jankowski) [inline src/bindings/js/node/include/type_validation.hpp:114]:
```suggestion
template <typename Arg, typename... Args>
std::vector<std::string> get_signature() {
    auto attributes = get_signature<Arg>();
    // pseudo code
    attributes.append(get_signature<Args...>());
    return attributes;
};
```

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/bindings/js/node/include/type_validation.hpp [hunk 2]
```diff
@@ -5,7 +5,142 @@
...
+namespace NapiArg {
+const char* get_type_name(napi_valuetype type);
+}  // namespace NapiArg
+
+namespace js {
+std::string get_current_signature(const Napi::CallbackInfo& info);
+
+template <typename T>
+const char* get_attr_type() {
+    throw std::runtime_error("get_attr_type 123 is not implemented for passed type!");
```
[the remark this answers is not in the document]
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/type_validation.hpp:43]: It leads to error:
image

[the remark this answers is not in the document]
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/type_validation.hpp]: Fixed

File: src/bindings/js/node/include/type_validation.hpp [hunk 3]
```diff
@@ -5,7 +5,142 @@
...
 
 #include "node/include/addon.hpp"
+#include "node/include/model_wrap.hpp"
+#include "node/include/tensor.hpp"
 #include "openvino/openvino.hpp"
+#include "openvino/util/common_util.hpp"
 
-/** @brief Checks if Napi::Value is a TensorWrap.*/
-bool is_tensor(const Napi::Env& env, const Napi::Value& value);
+namespace NapiTypename {
```
[the remark this answers is not in the document]
Comment (vishniakov-nikolai) [reply src/bindings/js/node/include/type_validation.hpp:15]: Fixed

File: src/bindings/js/node/include/type_validation.hpp [hunk 4]
```diff
@@ -5,7 +5,146 @@
...
+static const char EXTERNAL_STR[] = "external";
+static const char BIGINT_STR[] = "bigint";
+static const char UNKNOWN_STR[] = "unknown";
+}  // namespace NapiTypename
+
+namespace BindingTypename {
+static const char INT[] = "Integer";
+static const char MODEL[] = "Model";
+static const char TENSOR[] = "Tensor";
+static const char BUFFER[] = "Buffer";
```
Comment (t-jankowski) [inline src/bindings/js/node/include/type_validation.hpp:33]: Move the definition to cpp file. I think these are not needed in hpp.

File: src/bindings/js/node/include/type_validation.hpp [hunk 5]
```diff
@@ -5,7 +5,142 @@
...
+
+template <>
+const char* get_attr_type<ModelWrap>();
+
+template <>
+const char* get_attr_type<TensorWrap>();
+
+template <typename T>
+bool validate_value(const Napi::Env& env, const Napi::Value& arg) {
+    throw std::runtime_error("Validation for this type is not implemented!");
```
[the remark this answers is not in the document]
Comment (t-jankowski) [reply src/bindings/js/node/include/type_validation.hpp:67]: It looks `static_assert` doesn't work everywhere, then let's keep the runtime solution.

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/bindings/js/node/include/type_validation.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-jankowski) [reply src/bindings/js/node/include/type_validation.hpp]: So let's keep the runtime solution.

**Review by t-jankowski — APPROVED** [review]

File: src/bindings/js/node/include/type_validation.hpp [hunk 6]
```diff
@@ -5,7 +5,125 @@
...
+    return validate_impl<Ts...>(info, 0);
+};
+
+template <typename... Ts>
+bool validate(const Napi::CallbackInfo& info) {
+    return validate_detail<Ts...>(info);
+};
+
+template <typename... Ts>
+bool validate(const Napi::CallbackInfo& info, std::vector<std::string>& checked_signatures) {
```
Comment (praasz) [inline src/bindings/js/node/include/type_validation.hpp:120]:
```suggestion
bool validate(const Napi::CallbackInfo& info, std::vector<std::string>& allowed_signatures) {
```

File: src/bindings/js/node/src/core_wrap.cpp [hunk 7]
```diff
@@ -58,12 +60,36 @@ Napi::Function CoreWrap::get_class(Napi::Env env) {
...
+            model = _core.read_model(model_str, weight_tensor);
+        } else if (ov::js::validate<Napi::Buffer<uint8_t>>(info, checked_signatures)) {
+            std::string model_str = buffer_to_string(info[0]);
+            ov::Tensor weight_tensor = ov::Tensor(ov::element::Type_t::u8, {0});
+
+            model = _core.read_model(model_str, weight_tensor);
+        } else if (ov::js::validate<Napi::String>(info, checked_signatures)) {
+            model = _core.read_model(info[0].ToString());
+        } else {
+            std::string error_message = ov::js::get_parameters_error_msg(info, checked_signatures);
```
Comment (praasz) [inline src/bindings/js/node/src/core_wrap.cpp:89]:
```suggestion
            const auto error_message = ov::js::get_parameters_error_msg(info, checked_signatures);
```

File: src/bindings/js/node/include/type_validation.hpp [hunk 8]
```diff
@@ -5,7 +5,125 @@
...
+
+template <>
+const char* get_attr_type<ModelWrap>();
+
+template <>
+const char* get_attr_type<TensorWrap>();
+
+template <typename T>
+bool validate_value(const Napi::Env& env, const Napi::Value& arg) {
+    throw std::runtime_error("Validation for this type is not implemented!");
```
Comment (praasz) [inline src/bindings/js/node/include/type_validation.hpp:46]: Use OPENVINO_THROW ?

File: src/bindings/js/node/include/type_validation.hpp [hunk 9]
```diff
@@ -5,7 +5,125 @@
...
+namespace js {
+namespace NapiArg {
+const char* get_type_name(napi_valuetype type);
+}  // namespace NapiArg
+
+std::string get_current_signature(const Napi::CallbackInfo& info);
+
+template <typename T>
+const char* get_attr_type() {
+    throw std::runtime_error("get_attr_type is not implemented for passed type!");
```
Comment (praasz) [inline src/bindings/js/node/include/type_validation.hpp:23]:
```suggestion
    OPENVINO_THROW("get_attr_type is not implemented for passed type!");
```
?

**Review by praasz — APPROVED** [review]

**Comment by vishniakov-nikolai** [comment]
@praasz thank you for detailed review, I implemented the most suggestions that we've discussed

File: src/bindings/js/node/include/type_validation.hpp [hunk 10]
```diff
@@ -5,7 +5,101 @@
...
+
+template <typename... Ts>
+bool validate(const Napi::CallbackInfo& info) {
+    return InputParameters<Ts...>::validate(info, std::index_sequence_for<Ts...>{});
+};
+
+template <typename... Ts>
+bool validate(const Napi::CallbackInfo& info, std::vector<std::string>& allowed_signatures) {
+    const auto signature_attributes = get_signature<Ts...>();
+    allowed_signatures.push_back(signature_attributes);
```
Comment (praasz) [inline src/bindings/js/node/include/type_validation.hpp:98]:
```suggestion
    allowed_signatures.push_back(get_signature<Ts...>());
```
or 
```suggestion
    auto signature_attributes = get_signature<Ts...>();
    allowed_signatures.push_back(std::move(signature_attributes));
```

**Pull request merged**